xdelta3-dir-patcher

```diff
@@ -21,11 +56,735 @@ class AttributeDict(dict):
+        uid = lstat(absolute_path).st_uid
+        gid = lstat(absolute_path).st_gid
+        mode = S_IMODE(lstat(absolute_path).st_mode)
+
+        group = None
+        try:
+            group = getgrgid(gid)[0]
+        except KeyError as ke:
+            pass
+
+        user = None
+        try:
+            user = getpwuid(uid)[0],
+        except KeyError as ke:
+            pass
+
+        setter_func = getattr(dir_listing, method)
+
+        file_obj = setter_func(path.basename(absolute_path),
+                               absolute_path,
+                               mode,
+                               user,
+                               uid,
+                               group,
+                               gid,
+                               path.islink(absolute_path))
+
+        if file_obj and file_obj.is_link:
+            file_obj.link_target = readlink(absolute_path)
+
+        return file_obj
+
+    # XXX: Not thread safe when uninitialized
+    @property
+    def members(self):
+        if self._members:
+            return self._members
+
+        print('FS: Gathering filelist (%s/)' % path.basename(self.path))
+        member_tree = {}
+        member_tree['.'] = DirListing(path.basename(self.path))
+        for root, dirs, filenames in walk(self.path):
+            relative_path = path.relpath(root, self.path)
+
+            current_dir = DirListing()
+            member_tree[relative_path] = current_dir
+
+            parent_dir = path.dirname(relative_path)
+            if root != self.path and not parent_dir:
+                parent_dir = '.'
+
+            self._add_listing_object(current_dir, 'set_metadata', root)
+
+            if parent_dir in member_tree:
+                member_tree[parent_dir].add_subdir(current_dir)
+
+            files_to_process = filenames
+            for directory in dirs:
+                if path.islink(path.join(root, directory)):
+                    files_to_process.append(directory)
+
+            for filename in files_to_process:
+                absolute_path = path.join(root, filename)
+                relative_path = path.relpath(absolute_path, self.path)
+
+                # Have the listing for the file in the map but
+                # don't associate a DirListing object to it
+                file_dict = self._add_listing_object(current_dir,
+                                                     'add_file',
+                                                     absolute_path)
+
+                member_tree[relative_path] = file_dict
+
+        # Make sure that the root has a real unique value rather than '.'
+        member_tree[None] = member_tree.pop('.')
+
+        print('FS: Gathering completed (%s/)' % path.basename(self.path))
+
+        self._members = member_tree
+
+        return self._members
+
+    def _makedirs_with_check(self, directory):
+        if not path.lexists(directory):
+            makedirs(directory, exist_ok = True)
```

Comment:
@sgnn7 This doesn't make sense to me.

1. Why check for existence at all? Just run makedirs with exist_ok=True and skip the stating of the path.
2. If the path exists but isn't a directory, then this is an error that should be caught.